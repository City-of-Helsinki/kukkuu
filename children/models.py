from hashlib import sha256
from typing import Iterable, Optional

from django.conf import settings
from django.core.validators import RegexValidator
from django.db import models, transaction
from django.db.models import Q
from django.utils import timezone
from django.utils.translation import gettext_lazy as _

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from languages.models import Language
from users.models import Guardian


class ChildQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardians__user=user) | Q(project__in=user.administered_projects)
        ).distinct()

    def user_can_update(self, user):
        return self.filter(guardians__user=user)

    def user_can_delete(self, user):
        return self.filter(guardians__user=user)

    @transaction.atomic()
    def delete(self):
        for child in self:
            child.delete()


postal_code_validator = RegexValidator(
    regex=r"^\d{5}$", message="Postal code must be 5 digits", code="invalid_postal_code"
)


class Child(UUIDPrimaryKeyModel, TimestampedModel):
    name = models.CharField(verbose_name=_("name"), max_length=64, blank=True)
    birthyear = models.DateField(verbose_name=_("birthyear"))
    postal_code = models.CharField(
        verbose_name=_("postal code"),
        max_length=5,
        validators=[postal_code_validator],
    )
    guardians = models.ManyToManyField(
        Guardian,
        verbose_name=_("guardians"),
        related_name="children",
        through="children.Relationship",
        blank=True,
    )
    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="children",
        on_delete=models.PROTECT,
    )
    languages_spoken_at_home = models.ManyToManyField(
        Language,
        verbose_name=_("languages spoken at home"),
        related_name="children",
        blank=True,
    )

    objects = ChildQuerySet.as_manager()

    class Meta:
        verbose_name = _("child")
        verbose_name_plural = _("children")
        ordering = ["birthyear", "name"]

    def __str__(self):
        return f"{self.name} ({self.birthyear})"

    @transaction.atomic()
    def delete(self, *args, **kwargs):
        self.enrolments.upcoming().delete()
        return super().delete(*args, **kwargs)

    def can_user_administer(self, user):
        return user.can_administer_project(self.project)

    def get_enrolment_count(self, year: Optional[int] = None, past=False):
        from events.models import Event  # noqa

        if year and past:
            raise ValueError("Cannot use year and past arguments at the same time.")
        now = timezone.now()
        year = year or now.year

        occurrence_filters = Q(time__year=year)
        # For external ticket system events published_at field is used to determine the
        # event's year. That is obviously not a perfect solution, but the best we can
        # do with the current data, and most probably good enough.
        ticket_system_filters = Q(event__published_at__year=year)
        if past:
            occurrence_filters &= Q(time__lt=now)
            ticket_system_filters &= Q(event__published_at__lt=now)
        return (
            self.occurrences.filter(occurrence_filters).count()
            + self.ticket_system_passwords.filter(ticket_system_filters)
            .distinct()
            .count()
        )

    @staticmethod
    def _hash(values: Iterable[str]) -> str:
        return sha256(",".join(values).encode("utf8")).hexdigest()

    @property
    def _salt_birthyear_postal_code_guardian_emails_list(self) -> list[str]:
        """
        List of salt, child's birthyear, postal code and guardians' emails.
        """
        return [
            settings.KUKKUU_HASHID_SALT,
            self.birthyear.isoformat(),
            self.postal_code,
            *sorted(guardian.email for guardian in self.guardians.all()),
        ]

    @property
    def birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's birthyear, postal code and guardians' emails.
        """
        return self._hash(self._salt_birthyear_postal_code_guardian_emails_list)

    @property
    def name_birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's name, birthyear, postal
        code and guardians' emails.
        """
        return self._hash(
            [self.name] + self._salt_birthyear_postal_code_guardian_emails_list
        )


class RelationshipQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardian__user=user) | Q(child__project__in=user.administered_projects)
        ).distinct()


class Relationship(models.Model):
    PARENT = "parent"  # In Finnish: Vanhempi
    OTHER_GUARDIAN = "other_guardian"  # In Finnish: Muu huoltaja
    OTHER_RELATION = "other_relation"  # In Finnish: Muu l√§heinen
    ADVOCATE = "advocate"  # In Finnish: Virallinen puolesta-asioija
    TYPE_CHOICES = (
        (PARENT, _("Parent")),
        (OTHER_GUARDIAN, _("Other guardian")),
        (OTHER_RELATION, _("Other relation")),
        (ADVOCATE, _("Advocate")),
    )

    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    child = models.ForeignKey(
        Child,
        verbose_name=_("child"),
        related_name="relationships",
        on_delete=models.CASCADE,
    )
    guardian = models.ForeignKey(
        Guardian,
        verbose_name=_("guardian"),
        on_delete=models.CASCADE,
        related_name="relationships",
    )
    type = models.CharField(
        verbose_name=_("type"),
        choices=TYPE_CHOICES,
        max_length=64,
        null=True,
        blank=True,
    )

    objects = RelationshipQuerySet.as_manager()

    class Meta:
        verbose_name = _("relationship")
        verbose_name_plural = _("relationships")
        ordering = [
            "guardian",
            "child",
        ]
