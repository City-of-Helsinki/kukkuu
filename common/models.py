import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.managers import TranslatableQuerySet as ParlerTranslatableQuerySet
from parler.models import TranslatableModel as ParlerTranslatableModel


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True


class TranslatableQuerySet(ParlerTranslatableQuerySet):
    def create_translatable_object(self, **kwargs):
        translations = kwargs.pop("translations")
        obj = self.create(**kwargs)
        for translation in translations:
            language_code = translation.pop("language_code")
            obj.create_translation(language_code=language_code, **translation)
        return obj


class TranslatableModel(ParlerTranslatableModel):
    objects = TranslatableQuerySet.as_manager()

    class Meta:
        abstract = True
