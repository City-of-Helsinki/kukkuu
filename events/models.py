from django.conf import settings
from django.db import models
from django.db.models import Q
from django.utils import timezone
from django.utils.timezone import now
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatedFields

from children.models import Child
from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.consts import NotificationType
from events.utils import send_event_notifications_to_guardians
from venues.models import Venue


# This need to be inherited from TranslatableQuerySet instead of default model.QuerySet
class EventQueryset(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(project__users=user) | Q(published_at__isnull=False)
        ).distinct()

    def published(self):
        return self.filter(published_at__isnull=False)


class Event(TimestampedModel, TranslatableModel):
    CHILD_AND_GUARDIAN = "child_and_guardian"
    CHILD_AND_1_OR_2_GUARDIANS = "child_and_1_or_2_guardians"
    FAMILY = "family"
    PARTICIPANTS_PER_INVITE_CHOICES = (
        (CHILD_AND_GUARDIAN, _("Child and guardian")),
        (CHILD_AND_1_OR_2_GUARDIANS, _("Child and 1-2 guardians")),
        (FAMILY, _("Family")),
    )

    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255, blank=True),
        short_description=models.TextField(
            verbose_name=_("short description"), blank=True
        ),
        description=models.TextField(verbose_name=_("description"), blank=True),
        image_alt_text=models.CharField(
            verbose_name=_("image alt text"), blank=True, max_length=255
        ),
    )
    image = models.ImageField(blank=True, verbose_name=_("image"))
    participants_per_invite = models.CharField(
        max_length=255,
        choices=PARTICIPANTS_PER_INVITE_CHOICES,
        verbose_name=_("participants per invite"),
    )
    duration = models.PositiveSmallIntegerField(
        verbose_name=_("duration"), blank=True, null=True, help_text=_("In minutes")
    )
    capacity_per_occurrence = models.PositiveSmallIntegerField(
        verbose_name=_("capacity per occurrence")
    )
    published_at = models.DateTimeField(
        blank=True, null=True, verbose_name=_("published at")
    )

    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="events",
        on_delete=models.CASCADE,
    )

    objects = EventQueryset.as_manager()

    class Meta:
        verbose_name = _("event")
        verbose_name_plural = _("events")

    def __str__(self):
        name = self.safe_translation_getter("name", super().__str__())
        published_text = _("published") if self.published_at else _("unpublished")
        return f"{name} ({self.pk}) ({self.project.year}) ({published_text})"

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project_id).exists()

    def publish(self):
        self.published_at = timezone.now()
        self.save()

        send_event_notifications_to_guardians(
            self,
            NotificationType.EVENT_PUBLISHED,
            self.project.children.prefetch_related("guardians"),
        )

    def is_published(self):
        return bool(self.published_at)


class OccurrenceQueryset(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(event__project__users=user) | Q(event__published_at__isnull=False)
        ).distinct()

    def delete(self, *args, **kwargs):
        for obj in self:
            obj.delete()


class Occurrence(TimestampedModel):
    time = models.DateTimeField(verbose_name=_("time"))
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )
    venue = models.ForeignKey(
        Venue,
        verbose_name=_("venue"),
        related_name="occurrences",
        on_delete=models.CASCADE,
    )

    children = models.ManyToManyField(
        Child,
        verbose_name=_("children"),
        related_name="occurrences",
        through="events.Enrolment",
        blank=True,
    )
    occurrence_language = models.CharField(
        max_length=10,
        choices=settings.LANGUAGES,
        verbose_name=_("occurrence language"),
        default=settings.LANGUAGES[0][0],
    )
    objects = OccurrenceQueryset.as_manager()

    class Meta:
        verbose_name = _("occurrence")
        verbose_name_plural = _("occurrences")

    def __str__(self):
        return f"{self.time} ({self.pk})"

    def delete(self, *args, **kwargs):
        if self.time >= now():
            # this QS needs to be evaluated here, it would not work after the
            # occurrence has been deleted
            children = list(self.children.all())

            super().delete(*args, **kwargs)

            send_event_notifications_to_guardians(
                self.event,
                NotificationType.OCCURRENCE_CANCELLED,
                children,
                occurrence=self,
            )
        else:
            super().delete(*args, **kwargs)

    def get_enrolment_count(self):
        try:
            # try to use an annotated value
            return self.enrolment_count
        except AttributeError:
            return self.enrolments.count()

    def can_user_administer(self, user):
        # There shouldn't ever be a situation where event.project != venue.project
        # so we can just check one of them
        return user.projects.filter(pk=self.event.project.pk).exists()


class Enrolment(TimestampedModel):
    child = models.ForeignKey(
        Child,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("child"),
    )
    occurrence = models.ForeignKey(
        Occurrence,
        related_name="enrolments",
        on_delete=models.CASCADE,
        verbose_name=_("occurrence"),
    )
    attended = models.NullBooleanField(verbose_name=_("attended"))

    class Meta:
        verbose_name = _("enrolment")
        verbose_name_plural = _("enrolments")
        constraints = [
            models.UniqueConstraint(
                fields=["child", "occurrence"], name="unq_child_occurrence"
            )
        ]

    def __str__(self):
        return f"{self.pk} {self.child_id}"

    def save(self, *args, **kwargs):
        created = self.pk is None
        super().save(*args, **kwargs)

        if created:
            send_event_notifications_to_guardians(
                self.occurrence.event,
                NotificationType.OCCURRENCE_ENROLMENT,
                self.child,
                occurrence=self.occurrence,
            )

    def delete_and_send_notification(self):
        child = self.child
        occurrence = self.occurrence
        self.delete()

        send_event_notifications_to_guardians(
            occurrence.event,
            NotificationType.OCCURRENCE_UNENROLMENT,
            child,
            occurrence=occurrence,
        )

    def can_user_administer(self, user):
        return self.occurrence.can_user_administer(user)
