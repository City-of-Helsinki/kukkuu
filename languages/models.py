import gettext

import pycountry
from django.core.exceptions import ValidationError
from django.db import IntegrityError, models, transaction
from django.utils import translation
from django.utils.translation import ugettext_lazy as _
from parler.models import TranslatableModel, TranslatedFields

from common.models import TranslatableQuerySet
from kukkuu.settings import PARLER_SUPPORTED_LANGUAGE_CODES


class InvalidLanguageCodeError(Exception):
    pass


class LanguageAlreadyExistsError(Exception):
    pass


class LanguageQueryset(TranslatableQuerySet):
    @transaction.atomic
    def create_from_language_code(self, language_code):
        pycountry_language = get_pycountry_language(language_code)

        with translation.override("en"):
            try:
                obj = self.create(
                    alpha_3_code=pycountry_language.alpha_3,
                    name=pycountry_language.name,
                )
            except IntegrityError as e:
                raise LanguageAlreadyExistsError(e)

        for parler_language in PARLER_SUPPORTED_LANGUAGE_CODES:
            if parler_language == "en":
                continue

            gettext_translation = gettext.translation(
                "iso639-3", pycountry.LOCALES_DIR, languages=[parler_language]
            )
            obj.create_translation(
                parler_language,
                name=gettext_translation.gettext(pycountry_language.name),
            )

        return obj

    @transaction.atomic
    def create_option_other(self):
        obj, created = self.language("en").get_or_create(
            alpha_3_code=None,
            translations__name="Other language",
            defaults={"name": "Other language"},
        )
        if not created:
            raise LanguageAlreadyExistsError()

        obj.create_translation("sv", name="Annat spr√•k")
        obj.create_translation("fi", name="muu kieli")

        return obj


class Language(TranslatableModel):
    alpha_3_code = models.CharField(
        max_length=3,
        verbose_name=_("alpha-3 code"),
        help_text=_("ISO 639-3 (language) or ISO 639-5 (language family) alpha-3 code"),
        unique=True,
        null=True,
        blank=True,
    )
    translations = TranslatedFields(
        name=models.CharField(verbose_name=_("name"), max_length=255, blank=True),
    )

    objects = LanguageQueryset.as_manager()

    class Meta:
        verbose_name = _("language")
        verbose_name_plural = _("languages")

    def __str__(self):
        return f"{self.name} ({self.alpha_3_code})"

    def clean(self):
        if self.alpha_3_code:
            try:
                get_pycountry_language(self.alpha_3_code)
            except InvalidLanguageCodeError as e:
                raise ValidationError({"alpha_3_code": e})


def get_pycountry_language(language_code):
    language = pycountry.languages.get(
        alpha_3=language_code
    ) or pycountry.language_families.get(alpha_3=language_code)

    if not language:
        raise InvalidLanguageCodeError(f'Invalid language code "{language_code}"')

    return language
