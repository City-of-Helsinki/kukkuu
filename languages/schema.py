import graphene
from django.apps import apps
from graphene import relay
from graphene_django import DjangoConnectionField, DjangoObjectType
from languages.models import Language

LanguageTranslation = apps.get_model("languages", "LanguageTranslation")


class LanguageTranslationType(DjangoObjectType):
    class Meta:
        model = LanguageTranslation
        exclude = ("id", "master")


class LanguageNode(DjangoObjectType):
    name = graphene.String()

    class Meta:
        model = Language
        interfaces = (relay.Node,)
        fields = ("alpha_3_code", "name", "translations")

    @classmethod
    def get_queryset(cls, queryset, info):
        return queryset.prefetch_related("translations")


class Query:
    language = relay.Node.Field(LanguageNode)
    languages = DjangoConnectionField(LanguageNode)
