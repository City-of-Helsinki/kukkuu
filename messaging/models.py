from random import random

from django.db import models
from django.utils.timezone import now
from django.utils.translation import gettext_lazy as _
from parler.models import TranslatedFields
from projects.models import Project

from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.models import Event, Occurrence
from users.models import Guardian


class MessageQuerySet(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(project__users=user)


class Message(TimestampedModel, TranslatableModel):
    ALL = "all"
    INVITED = "invited"
    ENROLLED = "enrolled"
    ATTENDED = "attended"
    SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION = "subscribed_to_free_spot_notification"
    RECIPIENT_SELECTION_CHOICES = (
        (ALL, _("All")),
        (INVITED, _("Invited")),
        (ENROLLED, _("Enrolled")),
        (ATTENDED, _("Attended")),
        (
            SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION,
            _("Subscribed to free spot notification"),
        ),
    )

    project = models.ForeignKey(
        Project,
        verbose_name=_("project"),
        related_name="messages",
        on_delete=models.CASCADE,
    )
    translations = TranslatedFields(
        subject=models.CharField(verbose_name=_("subject"), max_length=255),
        body_text=models.TextField(verbose_name=_("body plain text")),
    )
    sent_at = models.DateTimeField(verbose_name=_("sent at"), blank=True, null=True)
    recipient_count = models.PositiveIntegerField(
        verbose_name=_("recipient count"), blank=True, null=True
    )
    recipient_selection = models.CharField(
        max_length=64,
        verbose_name=_("recipient selection"),
        choices=RECIPIENT_SELECTION_CHOICES,
    )
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="messages",
        blank=True,
        null=True,
        on_delete=models.CASCADE,
    )
    occurrences = models.ManyToManyField(
        Occurrence, verbose_name=_("occurrences"), related_name="messages", blank=True,
    )

    objects = MessageQuerySet.as_manager()

    class Meta:
        ordering = ("id",)
        verbose_name = _("message")
        verbose_name_plural = _("messages")

    def __str__(self):
        return f"({self.pk}) {self.subject} ({self.sent_at or 'not sent'})"

    def send(self):
        self.recipient_count = (
            Guardian.objects.filter(children__project=self.project).count()
            if self.recipient_selection == Message.ALL
            else random.randint(
                0, 50
            )  # TODO temporary thingy before actual send implementation
        )
        self.sent_at = now()
        self.save(update_fields=("recipient_count", "sent_at"))

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project.pk).exists()
