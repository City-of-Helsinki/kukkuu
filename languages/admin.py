from django.contrib import admin
from parler.admin import TranslatableAdmin

from .models import Language


@admin.register(Language)
class LanguageAdmin(TranslatableAdmin):
    list_display = ("alpha_3_code", "name")
