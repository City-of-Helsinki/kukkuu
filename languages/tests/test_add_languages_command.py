from io import StringIO

import pytest
from django.core.management import call_command

from kukkuu.settings import PARLER_SUPPORTED_LANGUAGE_CODES
from languages.models import Language


def call_command_and_assert_output(snapshot, *args, **kwargs):
    with StringIO() as out:
        call_command("add_languages", *args, **kwargs, stdout=out)
        snapshot.assert_match(out.getvalue())


def get_serialized_languages():
    ret = {}
    for language in Language.objects.all():
        name = {
            parler_language: language.safe_translation_getter(
                "name", language_code=parler_language
            )
            for parler_language in PARLER_SUPPORTED_LANGUAGE_CODES
        }
        ret[language.alpha_3_code] = name

    return ret


@pytest.fixture
def initial_language():
    return Language.objects.create(alpha_3_code="fin", name="INITIAL Finnish")


@pytest.mark.django_db
def test_no_arguments(snapshot):
    call_command_and_assert_output(snapshot)


@pytest.mark.django_db
def test_add_default_languages(snapshot):
    call_command_and_assert_output(snapshot, "--default")
    snapshot.assert_match(get_serialized_languages())


@pytest.mark.django_db
def test_add_languages(snapshot):
    call_command_and_assert_output(snapshot, "abd,agh")
    snapshot.assert_match(get_serialized_languages())


@pytest.mark.django_db
def test_add_languages_and_flush(snapshot, initial_language):
    call_command_and_assert_output(snapshot, "abd,agh", "--flush")
    snapshot.assert_match(get_serialized_languages())
