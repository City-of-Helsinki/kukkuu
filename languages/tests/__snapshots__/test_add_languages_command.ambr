# serializer version: 1
# name: test_add_default_languages
  '''
  Created special language "Other"
  Created Arabic (ara)
  Created Bengali (ben)
  Created German (deu)
  Created English (eng)
  Created Estonian (est)
  Created Persian (fas)
  Created Finnish (fin)
  Created French (fra)
  Created Hindi (hin)
  Created Italian (ita)
  Created Kurdish (kur)
  Created Nepali (macrolanguage) (nep)
  Created Norwegian (nor)
  Created Polish (pol)
  Created Portuguese (por)
  Created Romanian (ron)
  Created Russian (rus)
  Created Sami languages (smi)
  Created Somali (som)
  Created Spanish (spa)
  Created Albanian (sqi)
  Created Swedish (swe)
  Created Tagalog (tgl)
  Created Thai (tha)
  Created Turkish (tur)
  Created Ukrainian (ukr)
  Created Urdu (urd)
  Created Vietnamese (vie)
  Created Chinese (zho)
  All done!
  
  '''
# ---
# name: test_add_default_languages.1
  dict({
    'ara': dict({
      'en': 'Arabic',
      'fi': 'arabia',
      'sv': 'Arabiska',
    }),
    'ben': dict({
      'en': 'Bengali',
      'fi': 'bengali',
      'sv': 'Bengali',
    }),
    'deu': dict({
      'en': 'German',
      'fi': 'saksa',
      'sv': 'Tyska',
    }),
    'eng': dict({
      'en': 'English',
      'fi': 'englanti',
      'sv': 'Engelska',
    }),
    'est': dict({
      'en': 'Estonian',
      'fi': 'viro',
      'sv': 'Estniska',
    }),
    'fas': dict({
      'en': 'Persian',
      'fi': 'persia',
      'sv': 'Persiska',
    }),
    'fin': dict({
      'en': 'Finnish',
      'fi': 'suomi',
      'sv': 'Finska',
    }),
    'fra': dict({
      'en': 'French',
      'fi': 'ranska',
      'sv': 'Franska',
    }),
    'hin': dict({
      'en': 'Hindi',
      'fi': 'hindi',
      'sv': 'Hindi',
    }),
    'ita': dict({
      'en': 'Italian',
      'fi': 'italia',
      'sv': 'Italienska',
    }),
    'kur': dict({
      'en': 'Kurdish',
      'fi': 'kurdi',
      'sv': 'Kurdiska',
    }),
    'nep': dict({
      'en': 'Nepali (macrolanguage)',
      'fi': 'Nepali (macrolanguage)',
      'sv': 'Nepali',
    }),
    'nor': dict({
      'en': 'Norwegian',
      'fi': 'norja',
      'sv': 'Norska',
    }),
    'pol': dict({
      'en': 'Polish',
      'fi': 'puola',
      'sv': 'Polska',
    }),
    'por': dict({
      'en': 'Portuguese',
      'fi': 'portugali',
      'sv': 'Portugisiska',
    }),
    'ron': dict({
      'en': 'Romanian',
      'fi': 'romania',
      'sv': 'Rumänska',
    }),
    'rus': dict({
      'en': 'Russian',
      'fi': 'venäjä',
      'sv': 'Ryska',
    }),
    'smi': dict({
      'en': 'Sami languages',
      'fi': 'Sami languages',
      'sv': 'Sami languages',
    }),
    'som': dict({
      'en': 'Somali',
      'fi': 'somali',
      'sv': 'Somali',
    }),
    'spa': dict({
      'en': 'Spanish',
      'fi': 'espanja',
      'sv': 'Spanska',
    }),
    'sqi': dict({
      'en': 'Albanian',
      'fi': 'albania',
      'sv': 'Albanska',
    }),
    'swe': dict({
      'en': 'Swedish',
      'fi': 'ruotsi',
      'sv': 'Svenska',
    }),
    'tgl': dict({
      'en': 'Tagalog',
      'fi': 'tagalog',
      'sv': 'Tagalog',
    }),
    'tha': dict({
      'en': 'Thai',
      'fi': 'thai',
      'sv': 'Thailändska',
    }),
    'tur': dict({
      'en': 'Turkish',
      'fi': 'turkki',
      'sv': 'Turkiska',
    }),
    'ukr': dict({
      'en': 'Ukrainian',
      'fi': 'ukraina',
      'sv': 'Ukrainska',
    }),
    'urd': dict({
      'en': 'Urdu',
      'fi': 'urdu',
      'sv': 'Urdu',
    }),
    'vie': dict({
      'en': 'Vietnamese',
      'fi': 'vietnam',
      'sv': 'Vietnamesiska',
    }),
    'zho': dict({
      'en': 'Chinese',
      'fi': 'kiina',
      'sv': 'Kinesiska',
    }),
    None: dict({
      'en': 'Other language',
      'fi': 'muu kieli',
      'sv': 'Annat språk',
    }),
  })
# ---
# name: test_add_languages
  '''
  Created Manide (abd)
  Created Ngelima (agh)
  All done!
  
  '''
# ---
# name: test_add_languages.1
  dict({
    'abd': dict({
      'en': 'Manide',
      'fi': 'Manide',
      'sv': 'Manide',
    }),
    'agh': dict({
      'en': 'Ngelima',
      'fi': 'Ngelima',
      'sv': 'Ngelima',
    }),
  })
# ---
# name: test_add_languages_and_flush
  '''
  Flushed existing languages
  Created Manide (abd)
  Created Ngelima (agh)
  All done!
  
  '''
# ---
# name: test_add_languages_and_flush.1
  dict({
    'abd': dict({
      'en': 'Manide',
      'fi': 'Manide',
      'sv': 'Manide',
    }),
    'agh': dict({
      'en': 'Ngelima',
      'fi': 'Ngelima',
      'sv': 'Ngelima',
    }),
  })
# ---
# name: test_no_arguments
  '''
  Nothing to do. Hint: try --help.
  
  '''
# ---
